discovery-local/src/main/scala/akka/discovery/local/registration/LocalServiceRegistration.scala

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright (C) 2017 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.discovery.local.registration
+
+import java.nio.file.{ Files, Path }
+
+import akka.discovery.local.JsonFormat._
+import spray.json._
+
+import scala.collection.immutable.Seq
+import scala.io.Source
+
+class LocalServiceRegistration(serviceFile: Path) {
+
+  private val utf8 = "utf-8"
+
+  /**
+   * Checks if a <host>:<port> is already registered in the service file
+   * @param address the hostname or ip address
+   * @param port the port
+   * @return true if already registered else false
+   */
+  def isRegistered(address: String, port: Int): Boolean =
+    localServiceEntries.contains(LocalServiceEntry(address, port))
+
+  /**
+   * Adds a new <host>:<port> to the service file, if it isn't
+   * already registered.
+   * @param address the hostname or ip address
+   * @param port the port
+   */
+  def add(address: String, port: Int): Unit = {
+    if (!isRegistered(address, port))
+      writeToFile(localServiceEntries :+ LocalServiceEntry(address, port))
+  }
+
+  /**
+   * Removes an <host>:<port> from the service file if registered.
+   * @param address the hostname or ip address
+   * @param port the port
+   */
+  def remove(address: String, port: Int): Unit =
+    writeToFile(localServiceEntries.filterNot(_ == LocalServiceEntry(address, port)))
+
+  private def writeToFile(localServiceEntries: Seq[LocalServiceEntry]) =
+    Files.write(serviceFile, localServiceEntries.toJson.compactPrint.getBytes(utf8))
```

Comment:
You could get 2 discoveries attempting to write and corrupt the file this way. Needs to take filesystem lock https://docs.oracle.com/javase/7/docs/api/java/nio/channels/FileLock.html around these operations. This likely is good enough as we're not talking about any high performance stuff here, just avoiding writing into it from more apps